README
cropgui is a GTK GUI for lossless JPEG cropping

Of the pictures I'd like to put online, I've found that in 75% of the cases
where I want to retouch the photo, it's to crop it and nothing else. Since I
shoot in jpeg, it's a lossy process to load the jpeg in gimp, crop it, and
write the result.

It turns out that debian's jpegtran has a "-crop" flag which performs lossless
cropping of jpeg images as long as the crop is to a multiple of what the
manpage calls the "iMCU boundary", a (usually?) 8x8 block of pixels. This
feature may have been pioneered by Guido of jpegclub.org some years ago.

There's apparently a nice Windows front-end to this program, but I didn't find
a Linux one. So I wrote one! It's pretty basic, but it gets the job done. You
can download it below.

To run cropgui, either list files on the commandline or select them from a file
browser (in the latter case, you're returned to the browser after cropping the
selected file(s); hit 'cancel' to exit completely). The output filename is
chosen automatically, and never overwrites the original (but it will silently
overwrite an earlier cropped version). For example, if the input is "moon.jpg"
then the output is "moon-cropped.jpg".

Images are automatically scaled by a power of 2 (e.g., 1/2, 1/4 or 1/8) in
order to fit onscreen. After releasing the mouse button, the cropped image
boundary may move a little bit; this represents the limitation that the
upper-left corner must be at a multiple of 8x8 original image pixels.

cropgui is written in Python and requires python, python-tkinter,
python-imaging, python-imaging-tk, and libjpeg-progs. It is available under the
terms of the GNU GPL version 2 or later. 

cropgui's revision history is managed by git:
    https://github.com/jepler/cropgui

-- Jeff Epler
jepler@unpythonic.net


debian/changelog
cropgui (0.6~pre0) UNRELEASED; urgency=medium

  * Pre-tag the next release

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 19 Apr 2021 07:48:04 -0500

cropgui (0.5) buster; urgency=medium

  * Tag a new release to help packagers

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 19 Apr 2021 07:44:06 -0500

cropgui (0.3) stretch; urgency=medium

  * Tag a new release to help packagers
  * Use jpegexiforient instead of exiftool
  * Crop other types of images (may not be lossless)
  * Rotate other types of images via imagemagick convert
  * Just copy file to output if the cropped and uncropped dimensions match
  * Don't forcibly convert filenames to lowercase
  * Don't create output filenames like foo-crop-crop.jpg
  * Don't hardcode TIOCGWINSZ value, it's in module termios
  * Fix initial rotation of images with EXIF rotation
  * Make cropgui.desktop follow desktop file specification better
  * Work around gratuitious 'tobytes' breakage in pillow
  * Other bugfixes and enhancements
  * Update debian packaging

 -- Jeff Epler <jepler@unpythonic.net>  Sat, 30 Sep 2017 09:24:01 -0500

cropgui (0.2+git27ee40aeee) hardy; urgency=low

  * use absolute paths
  * don't apply scaling to jpegtrans dimensions

 -- Reuben <reuben_p@yahoo.com>  Thu, 30 Jul 2015 07:04:00 +1000

cropgui (0.2) hardy; urgency=low

  * remember directory for 'open' file choosers
  * handle corrupt or unreadable images better
  * allow interactive image rotation
  * use exif image orientation by default
  * prompt user for save location when it's unwritable 

 -- Jeff Epler <jepler@unpythonic.net>  Sun, 26 Jul 2009 12:18:52 -0500

cropgui (0.1.1) hardy; urgency=low

  * fix dash-compatibility of build script 

 -- Jeff Epler <jepler@unpythonic.net>  Sun, 26 Jul 2009 12:18:37 -0500

cropgui (0.1) unstable; urgency=low

  * Initial release.

 -- Jeff Epler <jepler@unpythonic.net>  Mon, 26 May 2008 21:49:30 -0500


log.py
#    a part of cropgui, a graphical front-end for lossless jpeg cropping
#    Copyright (C) 2009 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
import atexit
import os
import platform
import struct
import sys
import threading
try:
    import fcntl
    import termios
except:
    pass

lock = threading.RLock()

screen_width = screen_height = None
def screen_size():
    if not os.isatty(2): return 0, 0
    if platform.system() == 'Windows': return 80, 25
    res = fcntl.ioctl(2, termios.TIOCGWINSZ, "\0" * 4)
    return struct.unpack("hh", res)
screen_width, screen_height = screen_size()

last_width = 0

def locked(f):
    def fu(*args, **kw):
        lock.acquire()
        try:
            return f(*args, **kw)
        finally:
            lock.release()
    return fu

@locked
def progress(message, *args):
    if args: message = message % args
    global last_width
    if screen_width == 0: return
    message = message[:screen_width - 1]
    width = len(message)
    if width < last_width:
        message += " " * (last_width - width)
    sys.stderr.write(message + "\r")
    sys.stderr.flush()
    last_width = width

@locked
def log(message, *args):
    if args: message = message % args
    progress_clear()
    sys.stderr.write(message + "\n");
    sys.stderr.flush()

def progress_clear():
    if last_width: progress("")

atexit.register(progress_clear)


cropgtk.py
#!/usr/bin/python3
#    cropgui, a graphical front-end for lossless jpeg cropping
#    Copyright (C) 2009 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from cropgui_common import *
from cropgui_common import _

import gi
#from gi.repository import GObject as gobject
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk as gtk
from gi.repository import GLib
#import gtk.glade
from gi.repository import Gdk as gdk
import filechooser
from gi.repository import GdkPixbuf as GdkPixbuf

import sys
import traceback
import imghdr

# otherwise, on hardy the user is shown spurious "[application] closed
# unexpectedly" messages but denied the ability to actually "report [the]
# problem"
def excepthook(exc_type, exc_obj, exc_tb):
    try:
        w = app['window1']
    except NameError:
        w = None
    lines = traceback.format_exception(exc_type, exc_obj, exc_tb)
    print("".join(lines))
    m = gtk.MessageDialog(w,
                gtk.DialogFlags.MODAL | gtk.DialogFlags.DESTROY_WITH_PARENT,
                gtk.MessageType.ERROR, gtk.ButtonsType.OK,
                _("Stepconf encountered an error.  The following "
                "information may be useful in troubleshooting:\n\n")
                + "".join(lines))
    m.show()
    m.run()
    m.destroy()
sys.excepthook = excepthook

import cropgui_common
gladefile = os.path.join(os.path.dirname(cropgui_common.__file__),
                "cropgui.glade")

class DragManager(DragManagerBase):
    def __init__(self, g):
        self.g = g
        self.idle = None
        self.busy = False

        DragManagerBase.__init__(self)

        w = g['window1']
        i = g['eventbox1']

        i.connect('button-press-event', self.press)
        i.connect('motion-notify-event', self.motion)
        i.connect('button-release-event', self.release)
        w.connect('delete-event', self.close)
        w.connect('key-press-event', self.key)
        g['toolbutton1'].connect('clicked', self.done)
        g['toolbutton2'].connect('clicked', self.escape)
        g['toolbutton3'].connect('clicked', self.ccw)
        g['toolbutton4'].connect('clicked', self.cw)

    def ccw(self, event):
        self.rotate_ccw()
    def cw(self, event):
        self.rotate_cw()

    def coords(self, event):
        return event.x, event.y

    def press(self, w, event):
        if event.type == gdk.EventType._2BUTTON_PRESS:
            return self.done()
        x, y = self.coords(event)
        self.drag_start(x, y, event.state & gdk.ModifierType.SHIFT_MASK)

    def motion(self, w, event):
        x, y = self.coords(event)
        if event.state & gdk.ModifierType.BUTTON1_MASK:
            self.drag_continue(x, y)
        else:
            self.idle_motion(x, y)

    idle_cursor = gdk.Cursor(gdk.CursorType.WATCH)
    cursor_map = {
        DRAG_TL: gdk.Cursor(gdk.CursorType.TOP_LEFT_CORNER),
        DRAG_L: gdk.Cursor(gdk.CursorType.LEFT_SIDE),
        DRAG_BL: gdk.Cursor(gdk.CursorType.BOTTOM_LEFT_CORNER),
        DRAG_TR: gdk.Cursor(gdk.CursorType.TOP_RIGHT_CORNER),
        DRAG_R: gdk.Cursor(gdk.CursorType.RIGHT_SIDE),
        DRAG_BR: gdk.Cursor(gdk.CursorType.BOTTOM_RIGHT_CORNER),
        DRAG_T: gdk.Cursor(gdk.CursorType.TOP_SIDE),
        DRAG_B: gdk.Cursor(gdk.CursorType.BOTTOM_SIDE),
        DRAG_C: gdk.Cursor(gdk.CursorType.FLEUR)}

    def idle_motion(self, x, y):
        i = self.g['image1']
        if not i: return
        if self.busy: cursor = self.idle_cursor
        else:
            what = self.classify(x, y)
            cursor = self.cursor_map.get(what, None)
#        i.window.set_cursor(cursor)

    def release(self, w, event):
        x, y = self.coords(event)
        self.drag_end(x, y)

    def done(self, *args):
        self.result = 1
        self.loop.quit()

    def escape(self, *args):
        self.result = 0
        self.loop.quit()

    def save_and_stay(self, *args):
        self.result = 2
        self.loop.quit()

    def close(self, *args):
        self.result = -1
        self.loop.quit()

    def key(self, w, e):
        if e.keyval == gdk.KEY_Escape: self.escape()
        elif e.keyval == gdk.KEY_Return: self.done()
        elif e.string:
            if e.string == 'n': self.escape()
            elif e.string == 'q': self.close()
            elif e.string == 's': self.save_and_stay()
            elif e.string in ',<': self.rotate_ccw()
            elif e.string in '.>': self.rotate_cw()
        # Don't know whether other event handlers need it too, but if
        # this doesn't return True (True prevents further handlers from
        # being invoked), a return somehow double-triggers self.done(),
        # skipping the next image in a multi-file invocation.  Not clear
        # what's going on, but this stops it.
        return True

    def image_set(self):
        self.render()

    def render(self):
        if self.idle is None:
            self.idle = GLib.idle_add(self.do_render)

    def do_render(self):
        if not self.idle:
            return
        self.idle = None

        g = self.g
        i = g['image1']
        if not i:  # app shutting down
            return

        if self.image is None:
            pixbuf = GdkPixbuf.Pixbuf.new_from_data('\0\0\0',
                GdkPixbuf.Colorspace.RGB, 0, 8, 1, 1, 3)
            i.set_from_pixbuf(pixbuf)
            g['pos_left'].set_text('---')
            g['pos_right'].set_text('---')
            g['pos_top'].set_text('---')
            g['pos_bottom'].set_text('---')
            g['pos_width'].set_text('---')
            g['pos_height'].set_text('---')
            g['pos_ratio'].set_text('---')

        else:
            rendered = self.rendered()
            rendered = rendered.convert('RGB')
            i.set_size_request(*rendered.size)
            try:
                image_data = rendered.tostring()
            except:
                image_data = rendered.tobytes()
            pixbuf = GdkPixbuf.Pixbuf.new_from_data(image_data,
                GdkPixbuf.Colorspace.RGB, 0, 8,
                rendered.size[0], rendered.size[1], 3*rendered.size[0])

            tt, ll, rr, bb = self.get_corners()
            ratio = self.describe_ratio()

            g['pos_left'].set_text('%d' % ll)
            g['pos_right'].set_text('%d' % rr)
            g['pos_top'].set_text('%d' % tt)
            g['pos_bottom'].set_text('%d' % bb)
            g['pos_width'].set_text('%d' % (rr-ll))
            g['pos_height'].set_text('%d' % (bb-tt))
            g['pos_ratio'].set_text(self.describe_ratio())

        i.set_from_pixbuf(pixbuf)

        return False


    def wait(self):
        self.loop = GLib.MainLoop()
        self.result = -1
        self.loop.run()
        return self.result

display = gdk.Display().get_default()
wa = display.get_primary_monitor().get_workarea()
max_h = wa.height - 192
max_w = wa.width - 64

class App:
    def __init__(self):
        self.builder = gtk.Builder()
        self.builder.add_from_file(gladefile)
        #self.glade = gtk.glade.XML(gladefile)
        self.drag = DragManager(self)
        self.task = CropTask(self)
        self.dirchooser = None
        self['window1'].set_title(_("CropGTK"))

    def __getitem__(self, name):
        return self.builder.get_object(name)

    def log(self, msg):
        s = self['statusbar1']
        if s:
            s.pop(0)
            s.push(0, msg)
    progress = log

    def set_busy(self, is_busy=True):
        self.drag.busy = is_busy
        i = self['image1']
        if i:
            self.drag.idle_motion(*i.get_pointer())

    def run(self):
        drag = self.drag
        task = self.task

        prev_name = None
        for image_name in self.image_names():
            drag.save_prev_crop()
            self['window1'].set_title(
                _("%s - CropGTK") % os.path.basename(image_name))
            self.set_busy()
            try:
                image = Image.open(image_name)
                drag.round_x, drag.round_y = image_round(image)
                drag.w, drag.h = image.size
                scale = 1
                scale = max (scale, nextPowerOf2((drag.w-1)/(max_w+1)))
                scale = max (scale, nextPowerOf2((drag.h-1)/(max_h+1)))
                thumbnail = image.copy()
                thumbnail.thumbnail((drag.w//scale, drag.h//scale))
            except (IOError,) as detail:
                m = gtk.MessageDialog(self['window1'],
                    gtk.DialogFlags.MODAL | gtk.DialogFlags.DESTROY_WITH_PARENT,
                    gtk.MessageType.ERROR, gtk.ButtonsType.OK,
                    "Could not open %s: %s" % (image_name, detail))
                m.show()
                m.run()
                m.destroy()
                continue
            image_type = imghdr.what(image_name)
            drag.image = thumbnail
            drag.rotation = 1
            rotation = image_rotation(image)
            if rotation in (3,6,8):
                while drag.rotation != rotation:
                    drag.rotate_ccw()
            drag.scale = scale

            v = 2
            while v == 2:
                self.set_busy(0)
                v = self.drag.wait()
                self.set_busy()
                if v == -1: break      # user closed app
                if v == 0:
                    self.log("Skipped %s" % os.path.basename(image_name))
                    continue           # user hit "next" / escape
                if v == 2: # save but stick with this image
                    target = self.output_name(image_name,image_type,True,prev_name)
                    prev_name = target
                else:
                    target = self.output_name(image_name,image_type)
                if not target:
                    self.log("Skipped %s" % os.path.basename(image_name))
                    continue # user hit "cancel" on save dialog
                task.add(CropRequest(
                    image=image,
                    image_name=image_name,
                    corners=drag.get_corners(),
                    rotation=drag.rotation,
                    target=target,
                ))
            if v == -1: break # user closed app

    def image_names(self):
        if len(sys.argv) > 1:
            for i in sys.argv[1:]: yield i
        else:
            c = filechooser.Chooser(_("Select images to crop"), self['window1'])
            while 1:
                files = c.run()
                if not files: break
                for i in files: yield i

    def output_name(self, image_name, image_type, chooser=False, prev_name=None):
        image_name = os.path.abspath(image_name)
        i = os.path.basename(image_name)
        if chooser and prev_name is not None:
            d = os.path.dirname(prev_name)
            j = os.path.basename(prev_name)
        else:
            d = os.path.dirname(image_name)
            j = os.path.splitext(i)[0]
            if j.endswith('-crop'): j += os.path.splitext(i)[1]
            else: j += "-crop" + os.path.splitext(i)[1]
            if os.access(d, os.W_OK) and not chooser: return os.path.join(d, j)
        title = _('Save cropped version of %s') % i
        if self.dirchooser is None:
            self.dirchooser = filechooser.DirChooser(title, self['window1'])
        else:
            self.dirchooser.set_title(title)
        self.dirchooser.set_current_folder(d if os.access(d, os.W_OK) else desktop_name())
        self.dirchooser.set_current_name(j)
        r = self.dirchooser.run()
        if not r: return ''
        r = r[0]
        e = os.path.splitext(r)[1]
        if image_type == "jpeg":
            if e.lower() in ['.jpg', '.jpeg']: return r
            return e + ".jpg"
        elif e.lower() == "." + image_type: return r
        else: return e + "." + image_type

app = App()
try:
    app.run()
finally:
    app.task.done()
    del app.task
    del app.drag


cropgui_common.py
#    a part of cropgui, a graphical front-end for lossless jpeg cropping
#    Copyright (C) 2009 Jeff Epler <jepler@unpythonic.net>
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
from collections import namedtuple

from PIL import Image
from PIL import ImageFilter
from PIL import ImageDraw
import platform
import subprocess
import threading
import queue
import os
import math

def getoutput(c):
    p = subprocess.Popen(c, shell=isinstance(c, str), stdout=subprocess.PIPE)
    stdout, stderr = p.communicate()
    return stdout


def _(s): return s  # TODO: i18n

(
    DRAG_NONE,
    DRAG_TL, DRAG_T, DRAG_TR,
    DRAG_L,  DRAG_C, DRAG_R,
    DRAG_BL, DRAG_B, DRAG_BR
) = list(range(10))

def describe_ratio(a, b):
    if a == 0 or b == 0: return "degenerate"
    if a > b: return "%.2f:1" % (a*1./b)
    return "1:%.2f" % (b*1./a)

def clamp(value, low, high):
    if value < low: return low
    if high < value: return high
    return value

def nextPowerOf2(n):
    count = 0;
    ceiln = math.ceil(n)
    # First ceiln in the below condition is for the
    # case where ceiln is 0
    if (ceiln and not(ceiln & (ceiln - 1))):
        return ceiln

    while (ceiln != 0):
        ceiln >>= 1
        count += 1

    return 1 << count;


def get_cropspec(image, corners, rotation):
    t, l, r, b = corners
    w = r - l
    h = b - t
    # Technically these parameters should straightforwardly produce perfect
    # crops, but jpegtran is broken here in two regards: (1) it doesn't
    # recognise perfect rotated crops as such, so the '-perfect' switch
    # erroneously rejects correctly ICMU-aligned rotate-and-crop commands, and
    # (2) it mistakenly rounds out crops on rotated/flipped images so that even
    # the bottom right corner is ICMU-aligned.  Sigh.  Adding the 'f' suffixes
    # to the dimensions here at least solves (2), and seems to produce the same
    # results as you get by manually constructing a pipeline of '-perfect'
    # command lines.
    if image.format == "JPEG" or image.format == "MPO":
        return "%dfx%df+%d+%d" % (w, h, l, t)
    else:
        return "%dx%d+%d+%d" % (w, h, l, t)


def ncpus():
    if os.path.exists("/proc/cpuinfo"):
        return open("/proc/cpuinfo").read().count("bogomips") or 1
    return 1
ncpus = ncpus()


CropRequest = namedtuple(
    "CropRequest",
    ["image", "image_name", "corners", "rotation", "target"]
)


class CropTask(object):
    def __init__(self, log):
        self.log = log
        self.tasks = queue.Queue()
        self.threads = set(self.create_task() for i in range(ncpus))
        for t in self.threads: t.start()

    def done(self):
        for t in self.threads:
            self.tasks.put(None)
        for t in self.threads:
            t.join()

    def create_task(self):
        return threading.Thread(target=self.runner)

    def count(self):
        return len(self.tasks) + len(self.threads)

    def add(self, task):
        self.tasks.put(task)

    def runner(self):
        while 1:
            task = self.tasks.get()
            if task is None:
                break
            image = task.image
            image_name = task.image_name
            rotation_int = task.rotation
            target = task.target
            shortname = os.path.basename(target)
            self.log.progress(_("Cropping to %s") % shortname)

            t, l, r, b = task.corners
            cropspec = get_cropspec(image, task.corners, rotation_int)

            if rotation_int == 3:
                rotation = "180"
            elif rotation_int == 6:
                rotation = "90"
            elif rotation_int == 8:
                rotation = "270"
            else:
                rotation = "none"

            # Copy file if no cropping or rotation.
            if (r + b - l - t) == (image.width + image.height) and rotation == "none":
                command = ["nice", "cp", image_name, target]
                if platform.system() == 'Windows': command = ["copy", image_name, target]
            # JPEG crop uses jpegtran
            elif image.format == "JPEG" or image.format == "MPO":
                command = ["nice", "jpegtran"]
                if platform.system() == 'Windows': command = ["jpegtran"]
                if rotation != "none":
                    command += ["-rotate", rotation]
                command += [
                    "-copy", "all",
                    "-crop", cropspec,
                    "-outfile", target,
                    image_name,
                ]
            # All other images use ImageMagick convert.
            else:
                command = ["nice", "convert"]
                if platform.system() == 'Windows': command = ["magick"]
                if rotation != "none":
                    command += ["-rotate", rotation]
                command += [image_name, "-crop", cropspec, '+repage', target]

            print(" ".join(command))
            subprocess.call(command)
            subprocess.call(["exiftool", "-overwrite_original", "-Orientation=1", "-n", target])
            self.log.log(_("Cropped to %s") % shortname)

class DragManagerBase(object):
    def __init__(self):
        self.render_flag = 0
        self.show_handles = True
        self.state = DRAG_NONE
        self.round_x = None
        self.round_y = None
        self.image = None
        self.w = 0
        self.h = 0
        self.left = self.right = self.top = self.bottom = 0
        self.prev_w = self.prev_h = 0
        self.prev_top = self.prev_bottom = 0
        self.prev_left = self=prev_right = 0

    def save_prev_crop(self):
        self.prev_w = self.w
        self.prev_h = self.h
        self.prev_top = self.top
        self.prev_bottom = self.bottom
        self.prev_left = self.left
        self.prev_right = self.right

    def set_image(self, image):
        if image is None:
            # XXX Why are left,right,bottom deleted?
            if hasattr(self, 'left'): del self.left
            if hasattr(self, 'right'): del self.right
            if hasattr(self, 'bottom'): del self.bottom
            if hasattr(self, 'blurred'): del self.blurred
            if hasattr(self, 'xor'): del self.xor
            self._image = None
        else:
            self._orig_image = image.copy()
            self._rotation = 1
            self.image_or_rotation_changed()

    def apply_rotation(self, image):
        if self.rotation == 1: return image.copy()
        if self.rotation == 3: return image.transpose(Image.ROTATE_180)
        if self.rotation == 6: return image.transpose(Image.ROTATE_270)
        if self.rotation == 8: return image.transpose(Image.ROTATE_90)

    def image_or_rotation_changed(self):
        self._image = image = self.apply_rotation(self._orig_image)
        # If new image size matches previous image size then preserve previous crop rect
        if self.w == 0 or self.w != self.prev_w or self.h != self.prev_h:
            self.top, self.bottom = self.fix(0, self.h, self.h, self.round_y, self.rotation in (3, 8))
            self.left, self.right = self.fix(0, self.w, self.w, self.round_x, self.rotation in (3, 6))
        else:
            self.top, self.bottom = self.prev_top, self.prev_bottom
            self.left, self.right = self.prev_left, self.prev_right
        blurred = image.copy()
        mult = len(self.image.mode) # replicate filter for L, RGB, RGBA
        self.blurred = image.copy().filter(
            ImageFilter.SMOOTH_MORE).point([x/2 for x in range(256)] * mult)
        self.xor = image.copy().point([x ^ 128 for x in range(256)] * mult)
        self.image_set()
        self.render()

    def fix(self, a, b, lim, r, reverse):
        """
        a, b: interval to fix
        lim: upper bound
        r: rounding size
        reverse: True to treat the upper bound as the origin
        """
        a, b = sorted((int(a), int(b)))
        if reverse:
            a = lim - ((((lim - a) + r - 1) // r) * r)
            # Rotation of non-ICU-aligned images can push bounds outside
            # the visible area, and while it might be nice to be able to
            # expose this, I don't know how to; so, crop inwards
            # instead.
            if a < 0:
                a += r
        else:
            a = (a // r) * r
        return a, b

    def get_corners(self):
        return self.top, self.left, self.right, self.bottom

    def get_screencorners(self):
        t, l, r, b = self.get_corners()
        return(int(t/int(self.scale)), int(l/int(self.scale)),
               int(r/int(self.scale)), int(b/int(self.scale)))

    def describe_ratio(self):
        w = self.right - self.left
        h = self.bottom - self.top
        return describe_ratio(w, h)

    def set_stdsize(self, x, y):
        # if frame doesn't fit in image, scale, preserving aspect ratio
        if (x > self.w):
            y = y * self.w / x
            x = self.w
        if (y > self.h):
            x = x * self.h / y
            y = self.h

        # calculate new crop area, preserving center
        left = (self.left + self.right - x) / 2
        right = left + x
        top = (self.top + self.bottom - y) / 2
        bottom = top + y

        # move crop area into the image, if necessary
        if (left < 0):
            left = 0
            right = x
        if (right > self.w):
            right = self.w
            left = right - x
        if (top < 0):
            top = 0
            bottom = y
        if (bottom > self.h):
            bottom = self.h
            top = bottom - y

        self.set_crop (top, left, right, bottom)

    def set_crop(self, top, left, right, bottom):
        self.top, self.bottom = self.fix(top, bottom, self.h, self.round_y, self.rotation in (3, 8))
        self.left, self.right = self.fix(left, right, self.w, self.round_x, self.rotation in (3, 6))
        self.render()

    def get_image(self):
        return self._image
    image = property(get_image, set_image, None,
                "change the target of this DragManager")

    def rendered(self):
        if self.image is None: return None

        t, l, r, b = self.get_screencorners()

        assert isinstance(t, int), t
        assert isinstance(l, int), l
        assert isinstance(r, int), r
        assert isinstance(b, int), b

        mask = Image.new('1', self.image.size, 0)
        mask.paste(1, (l, t, r, b))
        image = Image.composite(self.image, self.blurred, mask)

        if self.show_handles:
            dx = (r - l) / 4
            dy = (b - t) / 4

            mask = Image.new('1', self.image.size, 1)
            draw = ImageDraw.Draw(mask)

            draw.line([l, t, r, t], fill=0)
            draw.line([l, b, r, b], fill=0)
            draw.line([l, t, l, b], fill=0)
            draw.line([r, t, r, b], fill=0)

            draw.line([l+dx, t, l+dx, t+dy, l, t+dy], fill=0)
            draw.line([r-dx, t, r-dx, t+dy, r, t+dy], fill=0)
            draw.line([l+dx, b, l+dx, b-dy, l, b-dy], fill=0)
            draw.line([r-dx, b, r-dx, b-dy, r, b-dy], fill=0)

            image = Image.composite(image, self.xor, mask)
        return image

    def classify(self, x, y):
        t, l, r, b = self.get_screencorners()
        dx = (r - l) / 4
        dy = (b - t) / 4

        if x < l: return DRAG_NONE
        if x > r: return DRAG_NONE
        if y < t: return DRAG_NONE
        if y > b: return DRAG_NONE

        if x < l+dx:
            if y < t+dy: return DRAG_TL
            if y < b-dy: return DRAG_L
            return DRAG_BL
        if x < r-dx:
            if y < t+dy: return DRAG_T
            if y < b-dy: return DRAG_C
            return DRAG_B
        else:
            if y < t+dy: return DRAG_TR
            if y < b-dy: return DRAG_R
            return DRAG_BR

    cursor_map = {
        DRAG_TL: 'top_left_corner',
        DRAG_L: 'left_side',
        DRAG_BL: 'bottom_left_corner',
        DRAG_TR: 'top_right_corner',
        DRAG_R: 'right_side',
        DRAG_BR: 'bottom_right_corner',
        DRAG_T: 'top_side',
        DRAG_B: 'bottom_side',
        DRAG_C: 'fleur'}

    def drag_start(self, x, y, fixed=False):
        self.x0 = x
        self.y0 = y
        self.t0 = self.top
        self.l0 = self.left
        self.r0 = self.right
        self.b0 = self.bottom
        self.state = self.classify(x, y)
        if self.state in (DRAG_TL, DRAG_TR, DRAG_BL, DRAG_BR):
            self.fixed_ratio = fixed
        else:
            # can't drag an edge and preserve ratio (what does that mean?)
            # dragging center always preserves ratio
            self.fixed_ratio = False

    def drag_continue(self, x, y):
        dx = (x - self.x0) * int(self.scale)
        dy = (y - self.y0) * int(self.scale)
        if self.fixed_ratio:
            ratio = (self.r0-self.l0) * 1. / (self.b0 - self.t0)
            if self.state in (DRAG_TR, DRAG_BL): ratio = -ratio
            if abs(dx/ratio) > abs(dy):
                dy = int(round(dx / ratio))
            else:
                dx = int(round(dy * ratio))
        new_top, new_left, new_right, new_bottom = self.get_corners()
        if self.state == DRAG_C:
            # A center drag bumps into the edges
            if dx > 0:
                dx = min(dx, self.w - self.r0)
            else:
                dx = max(dx, -self.l0)
            if dy > 0:
                dy = min(dy, self.h - self.b0)
            else:
                dy = max(dy, -self.t0)
        if self.state in (DRAG_TL, DRAG_T, DRAG_TR, DRAG_C):
            new_top = self.t0 + dy
        if self.state in (DRAG_TL, DRAG_L, DRAG_BL, DRAG_C):
            new_left = self.l0 + dx
        if self.state in (DRAG_TR, DRAG_R, DRAG_BR, DRAG_C):
            new_right = self.r0 + dx
        if self.state in (DRAG_BL, DRAG_B, DRAG_BR, DRAG_C):
            new_bottom = self.b0 + dy
        # A drag never moves left past right and so on
        if self.state != DRAG_C:
            new_top = min(self.bottom-1, new_top)
            new_left = min(self.right-1, new_left)
            new_right = max(self.left+1, new_right)
            new_bottom = max(self.top+1, new_bottom)

        self.set_crop(new_top, new_left, new_right, new_bottom)

    def drag_end(self, x, y):
        self.set_crop(self.top, self.left, self.right, self.bottom)
        self.state = DRAG_NONE

    def _flip_dimensions(self):
        self.w, self.h = self.h, self.w
        self.round_x, self.round_y = self.round_y, self.round_x

    def rotate_ccw(self):
        self._flip_dimensions()
        r = self.rotation
        if   r == 1: r = 8
        elif r == 8: r = 3
        elif r == 3: r = 6
        elif r == 6: r = 1
        self.rotation = r

    def rotate_cw(self):
        self._flip_dimensions()
        r = self.rotation
        if   r == 1: r = 6
        elif r == 6: r = 3
        elif r == 3: r = 8
        elif r == 8: r = 1
        self.rotation = r

    inverse = {1: 1, 3:3, 6:8, 8:6}

    def set_rotation(self, rotation):
        if rotation not in (1, 3, 6, 8):
            raise ValueError('Unsupported rotation %r' % rotation)

        print("rotation", self.rotation, "->", rotation)
        self._rotation = rotation
        self.image_or_rotation_changed()

    def get_rotation(self):
        return self._rotation
    rotation = property(get_rotation, set_rotation, None,
            'Set image rotation')

def image_rotation(i):
    if not hasattr(i, '_getexif'):
        print("no getexif?", type(i), getattr(i, '_getexif', None))
        return 1
    exif = i._getexif()
    if not isinstance(exif, dict):
        print("not dict?", repr(exif))
        return 1
    result = exif.get(0x112, None)
    print("image_rotation", result)
    return result or 1


def image_round(i):
    """Return (horizontal block size, vertical block size)"""
    if i.format == "JPEG" or i.format == "MPO":
        x = max(xsamp for _id, xsamp, ysamp, _qtable in i.layer)
        y = max(ysamp for _id, xsamp, ysamp, _qtable in i.layer)
        return x * 8, y * 8
    else:
        return 1, 1


_desktop_name = None
def desktop_name():
    global _desktop_name
    if not _desktop_name:
        _desktop_name = getoutput("""
            test -f ${XDG_CONFIG_HOME:-~/.config}/user-dirs.dirs && . ${XDG_CONFIG_HOME:-~/.config}/user-dirs.dirs
            echo -n ${XDG_DESKTOP_DIR:-$HOME/Desktop}""")
        if (not os.path.isdir(_desktop_name)
                or not os.access(_desktop_name, os.W_OK)):
            _desktop_name = os.expanduser("~")
    return _desktop_name

def output_name(image_name):
    dirname = os.path.dirname(image_name)
    basename = os.path.basename(image_name)
    if not os.access(dirname, os.W_OK):
        image_name = os.path.join(desktop_name(), basename)
    base, ext = os.path.splitext(image_name)
    target = base + "-crop.jpg"
    return target
